src/codegen/generate_grpc_stubs.py

```diff
@@ -0,0 +1,111 @@
+
+
+def is_relative_to(path: pathlib.PurePath, other: pathlib.PurePath) -> bool:
+    """Return whether or not this path is relative to the other path."""
+    try:
+        path.relative_to(other)
+        return True
+    except ValueError:
+        return False
+
+
+def remove_generated_files(stubs_path: pathlib.Path, proto_path: pathlib.Path):
+    """Remove files generated by protoc."""
+    for path in stubs_path.glob("*.py"):
+        path.unlink()
+
+    for path in stubs_path.iterdir():
+        if not is_relative_to(path, proto_path):
+            shutil.rmtree(path, ignore_errors=True)
+
+
+def generate_python_files(
+    stubs_path: pathlib.Path, proto_path: pathlib.Path, proto_files: Sequence[pathlib.Path]
+):
+    """Generate python files from .proto files with protoc."""
+    arguments = [
+        "protoc",
+        f"--proto_path={str(proto_path)}",
+        f"--proto_path={pkg_resources.resource_filename('grpc_tools', '_proto')}",
+        f"--python_out={str(stubs_path)}",
+        f"--mypy_out={str(stubs_path)}",
+        f"--grpc_python_out={str(stubs_path)}",
+    ]
+    arguments += [str(path.relative_to(proto_path)).replace("\\", "/") for path in proto_files]
+
+    print(proto_files)
+
+    grpc_tools.protoc.main(arguments)
+
+
+def fix_import_paths(
+    stubs_path: pathlib.Path, stubs_namespace: str, proto_parent_namespaces: Sequence[str]
+):
+    """Fix import paths of generated files."""
+    print("Fixing import paths")
+    grpc_codegened_file_paths = list(stubs_path.rglob("*pb2*py"))
+    imports_to_fix = [path.stem for path in grpc_codegened_file_paths if path.parent == stubs_path]
+    for path in grpc_codegened_file_paths:
+        print(f"Processing {path}")
+        data = path.read_bytes()
+        for name in imports_to_fix:
+            data = data.replace(
+                f"import {name}".encode(), f"from {stubs_namespace} import {name}".encode()
+            )
+
+        for namespace in proto_parent_namespaces:
+            data = data.replace(
+                f"from {namespace}".encode(),
+                f"from {stubs_namespace}.{namespace}".encode(),
+            )
+        path.write_bytes(data)
+
+
+def add_init_files(stubs_path: pathlib.Path, proto_path: pathlib.Path):
+    """Add __init__.py files to generated file directories."""
+    for dir in stubs_path.rglob(""):
+        if not is_relative_to(dir, proto_path) and dir.is_dir():
+            init_path = dir / "__init__.py"
+            print(f"Creating {init_path}")
+            init_path.write_bytes(b'"""Auto generated gRPC files."""\n')
+
+
+def blacken_code(stubs_path: pathlib.Path):
+    """Run black on generated files."""
+    print("Running black")
+    for py_path in itertools.chain(stubs_path.rglob("*.py"), stubs_path.rglob("*.pyi")):
+        if black.format_file_in_place(
+            src=py_path, fast=False, mode=Mode(line_length=100), write_back=black.WriteBack.YES
+        ):
+            print(f"reformatted {py_path}")
```

Comment:
> It runs protoc, fixes up imports, creates __init__.py files, and even runs black (however it's probably better to just exclude the protoc output in the black config in pyproject.toml).

Reformatting the stubs is slow and does not have much value. Let's delete this and exclude the stubs instead.